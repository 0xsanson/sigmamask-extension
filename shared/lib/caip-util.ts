export function isCaipChainId(caipChainId: string): boolean {
  return caipChainId.startsWith('eip155:');
}
export function getCaipChainIdFromEthChainId(ethChainId: string): string {
  const chainIdDecimal = ethChainId.startsWith('0x')
    ? parseInt(ethChainId, 16).toString(10)
    : ethChainId;
  return `eip155:${chainIdDecimal}`;
}

export function getEthChainIdDecFromCaipChainId(caipChainId: string): string {
  return caipChainId.replace(/^eip155:/, '')
}

export function getEthChainIdHexFromCaipChainId(caipChainId: string): string {
  const chainIdDecimal = getEthChainIdDecFromCaipChainId(caipChainId)
  return `0x${parseInt(chainIdDecimal, 10).toString(16)}`;
}

export function getEthChainIdNumFromCaipChainId(caipChainId: string): number {
  const chainIdDecimal = getEthChainIdDecFromCaipChainId(caipChainId)
  return parseInt(chainIdDecimal, 10)
}
