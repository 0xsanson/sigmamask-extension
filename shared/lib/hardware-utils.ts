import { isManifestV3 } from '../modules/mv3.utils';
import { LedgerTransportTypes } from '../constants/hardware-wallets';

export const getDefaultLedgerTransportType = (): LedgerTransportTypes => {
  if (isManifestV3) {
    // Then enforce the use of WebHID as the default, despite
    // the current context not supporting window.navigator.hid
    return LedgerTransportTypes.webhid;
  }

  return Object.hasOwnProperty.call(window.navigator, 'hid')
    ? LedgerTransportTypes.webhid
    : LedgerTransportTypes.u2f;
};

export const getRecommendedLedgerTransportType = (): LedgerTransportTypes => {
  // Currently the recommended ledger transport type is the default
  return getDefaultLedgerTransportType();
};

type LedgerTranslations = Record<LedgerTransportTypes, string>;
type LedgerTransportOption = {
  name: LedgerTranslations[LedgerTransportTypes];
  value: LedgerTransportTypes;
};

export const getLedgerTransportOption = (
  transportType: LedgerTransportTypes,
  translations: LedgerTranslations,
): LedgerTransportOption => {
  return {
    name: translations[transportType],
    value: transportType,
  };
};
