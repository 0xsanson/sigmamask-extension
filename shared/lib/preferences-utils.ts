import { isManifestV3 } from '../modules/mv3.utils';
import { LedgerTransportTypes } from '../constants/hardware-wallets';

export const getInitLedgerTransportType = (): string => {
  if (isManifestV3) {
    // Then enforce the use of WebHID as the default, despite
    // the current context not supporting window.navigator.hid
    return LedgerTransportTypes.webhid;
  }

  return Object.hasOwnProperty.call(window.navigator, 'hid')
    ? LedgerTransportTypes.webhid
    : LedgerTransportTypes.u2f;
};
