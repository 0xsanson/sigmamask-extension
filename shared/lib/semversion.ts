/**
 * A simple utility to parse and compare semver versions.
 */

import type ExtensionPlatform from '../../app/scripts/platforms/extension';

/**
 * A semver version object.
 */
export type SemVersion = {
  major: number;
  minor: number;
  patch: number;
};

/**
 * Parse a semver version string into a SemVersion object.
 *
 * @param version - The version string to parse, of the form "major.minor.patch". e.g. "1.2.3"
 * @returns
 */
export const parseVersion = (version: string = ''): SemVersion | null => {
  const [major, minor, patch] = version.split('.').map(Number);

  // if version is not a valid semver, return false
  if ([major, minor, patch].some((num) => isNaN(num) || num < 0)) {
    return null;
  }
  return { major, minor, patch };
};

/**
 * Compare two SemVersion objects.
 *
 * @param a
 * @param b
 * @returns negative if a < b, zero if a === b, positive if a > b
 */
export const compareVersions = (a: SemVersion, b: SemVersion): number => {
  if (a.major !== b.major) {
    return a.major - b.major;
  }
  if (a.minor !== b.minor) {
    return a.minor - b.minor;
  }
  return a.patch - b.patch;
};

/**
 * Get the version of the extension.
 */
export const getMetamaskVersion = () =>
  parseVersion((global.platform as ExtensionPlatform).getVersion());
