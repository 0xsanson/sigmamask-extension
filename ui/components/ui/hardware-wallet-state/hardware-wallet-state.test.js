import React from 'react';
import { act } from '@testing-library/react-hooks';
import configureMockStore from 'redux-mock-store';
import { renderWithProvider } from '../../../../test/lib/render-helpers';
import mockState from '../../../../test/data/mock-state.json';
import HardwareWalletState from '.';

describe('HardwareWalletState Component', () => {
  const mockStore = configureMockStore()(mockState);

  it('should match default snapshot', () => {
    const { container } = renderWithProvider(
      <HardwareWalletState />,
      mockStore,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render message with locked state', () => {
    const props = {
      initialStatus: 'locked',
    };

    const { container } = renderWithProvider(
      <HardwareWalletState {...props} />,
      mockStore,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render empty with unlocked state', () => {
    const props = {
      initialStatus: 'unlocked',
    };

    const { container } = renderWithProvider(
      <HardwareWalletState {...props} />,
      mockStore,
    );

    expect(container).toMatchSnapshot();
  });

  it('should call onUpdate handler', async () => {
    let status = 'unlocked';
    const handler = (state) => {
      status = state;
    };
    // start unlocked and wait to update to locked
    const props = {
      initialStatus: status,
      pollingRateMs: 600,
      onUpdate: handler,
    };
    const { container } = renderWithProvider(
      <HardwareWalletState {...props} />,
      mockStore,
    );
    // should not change until polling commences
    expect(status).toStrictEqual('unlocked');
    // wait for polling to complete (allow time for at least 2 polling cycles)
    await act(() => new Promise((res) => setTimeout(res, 1500)));
    expect(status).toStrictEqual('locked');
    expect(container).toMatchSnapshot();
  });
});
