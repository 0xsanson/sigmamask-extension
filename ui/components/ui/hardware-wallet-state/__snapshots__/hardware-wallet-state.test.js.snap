// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HardwareWalletState Component should call onUpdate handler 1`] = `
<div>
  <div
    class="box mm-banner-base mm-banner-alert mm-banner-alert--severity-danger box--padding-3 box--padding-left-2 box--display-flex box--gap-2 box--flex-direction-row box--background-color-error-muted box--rounded-sm"
  >
    <div
      class="box mm-icon mm-icon--size-lg box--flex-direction-row box--color-error-default"
      style="mask-image: url('./images/icons/danger.svg');"
    />
    <div>
      <p
        class="box mm-text mm-text--body-md mm-text--color-text-default box--flex-direction-row"
      >
        Cannot connect to Ledger device. Please make sure your device is unlocked and Ethereum app is opened.
      </p>
    </div>
  </div>
</div>
`;

exports[`HardwareWalletState Component should match default snapshot 1`] = `
<div>
  <div
    class="box mm-banner-base mm-banner-alert mm-banner-alert--severity-danger box--padding-3 box--padding-left-2 box--display-flex box--gap-2 box--flex-direction-row box--background-color-error-muted box--rounded-sm"
  >
    <div
      class="box mm-icon mm-icon--size-lg box--flex-direction-row box--color-error-default"
      style="mask-image: url('./images/icons/danger.svg');"
    />
    <div>
      <p
        class="box mm-text mm-text--body-md mm-text--color-text-default box--flex-direction-row"
      >
        Cannot connect to Ledger device. Please make sure your device is unlocked and Ethereum app is opened.
      </p>
    </div>
  </div>
</div>
`;

exports[`HardwareWalletState Component should render empty with unlocked state 1`] = `<div />`;

exports[`HardwareWalletState Component should render message with locked state 1`] = `
<div>
  <div
    class="box mm-banner-base mm-banner-alert mm-banner-alert--severity-danger box--padding-3 box--padding-left-2 box--display-flex box--gap-2 box--flex-direction-row box--background-color-error-muted box--rounded-sm"
  >
    <div
      class="box mm-icon mm-icon--size-lg box--flex-direction-row box--color-error-default"
      style="mask-image: url('./images/icons/danger.svg');"
    />
    <div>
      <p
        class="box mm-text mm-text--body-md mm-text--color-text-default box--flex-direction-row"
      >
        Cannot connect to Ledger device. Please make sure your device is unlocked and Ethereum app is opened.
      </p>
    </div>
  </div>
</div>
`;
